presto-spark-base/src/main/java/com/facebook/presto/spark/execution/NativeExecutionTask.java

```diff
@@ -15,259 +15,196 @@
         requireNonNull(httpClient, "httpClient is null");
         requireNonNull(tableWriteInfo, "tableWriteInfo is null");
-        requireNonNull(stats, "stats is null");
         requireNonNull(executor, "executor is null");
         requireNonNull(updateScheduledExecutor, "updateScheduledExecutor is null");
-        requireNonNull(taskUpdateRequestCodec, "taskUpdateRequestCodec is null");
-        requireNonNull(planFragmentCodec, "planFragmentCodec is null");
-        requireNonNull(taskInfoCodec, "taskInfoCodec is null");
 
         try (SetThreadName ignored = new SetThreadName("NativeExecutionTask-%s", taskId)) {
             this.taskId = taskId;
             this.session = session;
             this.planFragment = planFragment;
-            this.httpClient = httpClient;
             this.tableWriteInfo = tableWriteInfo;
-            this.stats = stats;
             this.sources = sources;
             this.executor = executor;
             this.updateScheduledExecutor = updateScheduledExecutor;
             this.outputBuffers = createInitialEmptyOutputBuffers(PARTITIONED);
-            List<BufferInfo> bufferStates = outputBuffers.getBuffers()
-                    .keySet().stream()
-                    .map(outputId -> new BufferInfo(outputId, false, 0, 0, PageBufferInfo.empty()))
-                    .collect(toImmutableList());
-            this.taskInfo.set(createInitialTask(taskId, URI.create(LOCAL_URI + taskId), bufferStates, new TaskStats(DateTime.now(), null), ""));
-            this.taskUpdateRequestCodec = taskUpdateRequestCodec;
-            this.planFragmentCodec = planFragmentCodec;
-            this.taskInfoCodec = taskInfoCodec;
+            this.workerClient = new PrestoSparkHttpWorkerClient(httpClient, taskId, location);
+            this.taskInfoFetcher = new HttpNativeExecutionTaskInfoFetcher(
+                    this.updateScheduledExecutor,
+                    this.workerClient,
+                    this.executor,
+                    this.taskId);
+            this.taskResultFetcher = new HttpNativeExecutionTaskResultFetcher(
+                    this.updateScheduledExecutor,
+                    this.workerClient,
+                    this.taskId,
+                    Optional.empty());
         }
     }
 
+    /**
+     * Gets the most updated TaskInfo of the task of the native task.
+     *
+     * @return most updated TaskInfo, null if TaskInfoFetcher has not yet retrieved the very first TaskInfo.
+     */
     public TaskInfo getTaskInfo()
     {
-        return taskInfo.get();
+        return taskInfoFetcher.getTaskInfo();
     }
 
-    public void start()
+    /**
+     * Blocking call to poll from result fetcher buffer. Blocks until content becomes available in the buffer, or until timeout is hit.
+     *
+     * @return an Optional of the first SerializedPage result fetcher buffer contains, an empty Optional if no result is in the buffer.
+     */
+    public Optional<SerializedPage> pollResult()
```

Comment:
Can we have unit test coverage?

Reply:
Will come in the following commits, under development already